### DISCUSSION TITLE
remote debugging and mod loading over mqtt (or ws)

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1254] [Ideas]
Most of my devices are experimental and in the field. As such they're basically always in debug mode, or ready to have a debugger connected. I'd like to explore how to best connect them to a "management and debugging server/console" over the network. The idea would be that some kernel runs on the device and manages an always-active connection to a server. This connection would be leveraged by logging, debug, and application messages.

I'm considering several options:

## C-based kernel

For esp32 devices there is a lot of functionality available in esp-idf. The device could have a thread implemented in C that uses the esp-idf implementations of MQTT or WS to connect to the server. The C code would act as a host to XS, i.e., essentially be a beefed-up version of the current xsHost/xsPlatform. There is already code to run the debug protocol over TCP which could perhaps be leveraged.

The C code would provide MQTT abstractions to XS such that mqtt.js or similar could be tweaked to call into the C-managed MQTT connection instead of using the JavaScript MQTTclient, etc. A similar replacement of the WiFi modules could be provided to at least provide visibility into the WiFi connectivity.

The pros of this approach are:
- fits into the notion of a host for XS
- debug is implemented at a low level and would be always available if the device powers up and can connect
- none of this is visible at the JS level, except that some functionality (e.g. MQTT) is magically provided by the host

The cons are:
- the implementation would be esp-idf specific
- this puts the main server connection into C-land with all the programming/debugging fun that entails
- managing memory can become a problem given that any networking and TLS in particular are pretty hungry and this approach is bound to result in fragmentation between the C heap and XS memory

## Kernel as main JS VM

The connection to the server could perhaps also be implemented in the JS main VM. The JS kernel code would start running without debugger connection, connect to wifi and via MQTT to the server, and then start the JS app in a worker VM. It would set-up the debugger connection for the worker VM over MQTT.

Something that is confusing to me is that platform functions like `fxReceive()`, `fxConnect()`, etc. have a `txMachine` argument, which makes them scoped to a VM yet there is only one serial connection. Is there a logical debug connection per VM and if so where is the multiplexing done.

In my use-case what really matters is that the main VM can run without having a debug connection and that worker VMs can have debug connections. Dunno whether that's possible.

The pros of this approach are:
- as much as possible is in JS making it more portable and less magic, and fitting better with the ecma-419 networking stack
- implementing everything and customizing how things work would be easier

The cons are:
- more JS code needs to run successfully in order to be able to run the debug protocol
- running the app in a worker may end up causing frustrating limitations (the VMs could perhaps be flipped with the kernel running in a worker)

## Logging-only implementation

Instead of implementing the xs debug protocol the host portion could hook into `trace()` and exception handling and only provide logging functionality. This could be augmented with functionality to load/manage mods and, memory permitting, to run `eval()` on a remotely sent string. (This is essentially what I've been using for several years with Micropython.)

The pros of this approach:
- don't have to deal with the undocumented XS debug protocol and commands
- potentially much simpler overall

The cons:
- doesn't provide debugging functionality

## xsbug issues

One other issue is that xsbug currently can only talk to one board. With the approach proposed here there could be many boards on the network that all have a debug connection. This raises a number of questions:
- an intermediary is needed to connect one of the boards' debug connection to a user's xsbug at a time, i.e., the user would pick which board to connect to from some menu and that would set-up a bridge
- alternatively, xsbug could be taught to connect to a board instead of having boards spontaneously connect to xsbug when they start
- an overall question is what assumptions are build into the debug connection and xsbug: does the debug connection have to be there when the VM boots or can it be established much later?

The last point is perhaps one of the central ones: what would be needed to allow xsbug to connect to a board that has been running for a while?

**Comment by phoddie** [comment]
Lots of stuff here. Approaches that get deep into ESP32 specifics are less interesting to me. XS debugging support is flexible enough to do what you want, and do so in a way that is light and portable. 

I'm just going to share some information about the debugging protocol and xsbug. That should help correct some misunderstandings and, I hope, set you on a path towards a solution.

xsbug can support any number of virtual machines. You've seen this when debugging works on ESP32.

Each virtual machine connects to xsbug with its own TCP socket.

When debugging a device over serial (normal ESP32 case), `serial2xsbug` bridges serial to TCP. All VMs on the ESP32 share a single serial connection. serial2xsbug takes care of opening a TCP socket to xsbug for each virtual machine and routing messages to/from the appropriate virtual machine.

In the simulator, each virtual machine connects directly to xsbug using a TCP socket.

There is support in the ESP32 host for connecting to xsbug over Wi-Fi. It works, though it isn't used often so it may need a little maintenance. (Fun fact: ESP8266 supported debugging over Wi-Fi before debugging over serial)

A virtual machine can connect to and disconnect from the debugger at any time. The default behavior is to do this at start-up. Here's a small program that connects/disconnects every 5 seconds.

**main.js**
```js
import Timer from "timer";

class Debug {
	static isConnected() @ "xs_debug_isConnected"
	static connect() @ "xs_debug_connect"
	static disconnect() @ "xs_debug_disconnect"
}

let count = 0;
Timer.repeat(() => {
	if (Debug.isConnected()) {
		trace(`goodbye #${++count}\n`);
		Debug.disconnect();
	}
	else {
		Debug.connect();
		trace(`hello #${++count}\n`);
	}
}, 5000);
```

** debugmethods.c**
```c
#include "mc.xs.h"			// for xsID_ values


extern xsBooleanValue fxIsConnected(xsMachine* the);
extern xsBooleanValue fxLogin(xsMachine* the);
extern xsBooleanValue fxLogout(xsMachine* the);

void xs_debug_isConnected(xsMachine *the)
{
	xsmcSetBoolean(xsResult, fxIsConnected(the));
}

void xs_debug_connect(xsMachine *the)
{
	fxLogin(the);
}

void xs_debug_disconnect(xsMachine *the)
{
	fxLogout(the);
}
```

**manifest.json**
```json
{
	"include": "$(MODDABLE)/examples/manifest_base.json",
	"modules": {
		"*": [
			"./main",
			"./debugmethods"
		]
	}
}
```

If you have multiple ESP32 devices connected to your computer, you can debug those at the same time with xsbug. You'll have to launch serial2xsbug for each in separate console sessions.

**console session 1**
```
serial2xsbug /dev/device_1_port 460800 8N1
```

**console session 2**
```
serial2xsbug /dev/device_2_port 460800 8N1
```

etc.

Since xsbug connects with TCP sockets, you can also have serial2xsbug connect to xsbug running on another machine:

```
XSBUG_HOST=10.0.1.30 serial2xsbug /dev/device_port 460800 8N1
XSBUG_HOST=xsbug.tve.org serial2xsbug /dev/device_port 460800 8N1
```

For your scenario, assuming the Wi-Fi debugging support on-device works, you could have MQTT send a command to connect to a debugger running on a specified IP address.

**Reply by tve** [reply]
Ah, thanks, this is very helpful! I agree that an XS/JS-level solution is more interesting than esp-idf. I got thrown off about the multiple machines connecting to xsbug by mcconfig. I ran two of those and I believe the makefile kills all other connections. Or maybe I made a mistake.

All this seems to indicate that the simplest set-up might be something like this:
- Single VM
- Listen to debug "management command" topic to control when to connect/disconnect debugging
- Call fxLogin/fxLogout as appropriate
- Optionally split off logging (`trace`) and send via MQTT so it's always available whether 
- Use some config var to enable auto-connect of debugging at start
- Possibly detect if serial is connected at boot and run serial debug protocol instead (also fall-back to serial if no wifi/mqtt connection info is stored)

The complement to all this would be:
- CLI to tell a board to connect/disconnect xsbug
- web page served up by mosquitto to display available boards, view their logs, and tell them to connect/disconnect to xsbug
- Bonus points for figuring out how to embed xsbug into that web page...

The main issue with a single VM is that breakpoints are terminal since (I assume) the whole VM is stopped. So I really need one VM for the debug link and another for the app.

**Reply by phoddie** [reply]
> I ran two of those and I believe the makefile kills all other connections

Yes. The makefile enforces "one at a time" use of `serial2xsbug` because that's the right thing 99.9% of the time.

> The main issue with a single VM is that breakpoints are terminal since (I assume) the whole VM is stopped

Yes: breakpoint are breakpoints. They stop the VM. ;)

**Comment by tve** [comment]
I've been pondering things and browsing a lot of code... The big issues I see are flow control, halting the VM, and memory usage.

- if there's one MQTT connection to "the" server then flow control is a problem especially when the application VM is halted on a breakpoint: the connection needs to keep flowing for debug yet the app is obviously not consuming any messages; even when not halted the app could be very slow in consuming messages which then causes debug/logging timeouts
- if separate MQTT connections are used then memory consumption can become an issue and MQTTS becomes unpalatable in many applications

I'm wondering whether a more practical solution would be to just provide remote logging and perhaps `eval`. Maybe this could be extended to have 'eval-at' functionality where one can click on a line like setting a breakpoint and enter a code snippet that gets evaluated when the line is reached without halting the VM per-se. Then complement all this with having a couple rPi-Zero or esp32-s2 boards on hand that can run serial2xsbug and that one plugs in physically to get the full remote debugging when the simple stuff fails...

---

On the how-to front I'm a little stuck about how to engineer hooking into the logging and/or debug connection...

Starting with logging, it seems that I need to hook/replace `fxReport`, `fxReportException`, `fxReportWarning`, and `fxReportError`; perhaps more. All of those are in `xsDebug.c` so not easy to replace. I don't know what `fxBubble` does and whether that's also for "logging" in the general sense. One option could be to have a `#define extLogging` that is set in the manifest and that causes these functions to be omitted from `xsDebug.c` so they can be implemented externally.

An alternative could be to have a #define that causes `fxReport` et. al. to use the console logging code in addition to `fxVReport` and then to just hook c_printf. Maybe just undefining `mxNoConsole` for debug builds achieves that.

At a lower level, to hook into the debug protocol I essentially have to duplicate the whole framing and polling machinery in `esp/xsPlatform.c` and I'm not sure how to do that cleanly. Perhaps add a `kExternalConnection` next to `kSerialConnection` and/or `isExternalIP` next to `isSerialIP` (`xsPlatform.c` lines 67-68) and then add clauses based on that, for example, `fxConnect` could have:
```
if (isSerialIP(gXSBUG)) {
  ...
} else if (isExternalIP(gXSBUG)) {
  extFxConnect(the); // call external hook
} else {
  ...
}
```

I'm also not clear about a number of things related to the debug connection:
- `fxReceiveLoop` has nothing about a TCP debug connection, why is that not needed?
- when is `doRemoteCommand` invoked? It seems to be in cases where `doCommand` doesn't process a command and leaves it in `debugFragments`?

**Reply by phoddie** [reply]
I don't understand the first part, so I'm going to stick to that.

> The big issues I see are flow control, halting the VM, and memory usage.

OK.

It seems most of this gets better by using a separate worker for the one MQTT connection. That worker will trigger the breakpoint in the app VM. It will stay alive while the app is stopped at a breakpoint, so can do whatever you like to manage the flow control.

> the connection needs to keep flowing for debug

I don't follow this.
